pystac/asset.py

```diff
@@ -0,0 +1,165 @@
+from typing import Any, Dict, List, Optional, TYPE_CHECKING, Union
+
+import pystac as ps
+from pystac.utils import is_absolute_href, make_absolute_href
+
+if TYPE_CHECKING:
+    from pystac.collection import Collection as Collection_Type
+    from pystac.item import Item as Item_Type
+
+
+class Asset:
+    """An object that contains a link to data associated with an Item or Collection that
+    can be downloaded or streamed.
+
+    Args:
+        href (str): Link to the asset object. Relative and absolute links are both
+            allowed.
+        title (str): Optional displayed title for clients and users.
+        description (str): A description of the Asset providing additional details,
+            such as how it was processed or created. CommonMark 0.29 syntax MAY be used
+            for rich text representation.
+        media_type (str): Optional description of the media type. Registered Media Types
+            are preferred. See :class:`~pystac.MediaType` for common media types.
+        roles ([str]): Optional, Semantic roles (i.e. thumbnail, overview,
+            data, metadata) of the asset.
+        properties (dict): Optional, additional properties for this asset. This is used
+            by extensions as a way to serialize and deserialize properties on asset
+            object JSON.
+
+    Attributes:
+        href (str): Link to the asset object. Relative and absolute links are both
+            allowed.
+        title (str): Optional displayed title for clients and users.
+        description (str): A description of the Asset providing additional details,
+            such as how it was processed or created. CommonMark 0.29 syntax MAY be
+            used for rich text representation.
+        media_type (str): Optional description of the media type. Registered Media Types
+            are preferred. See :class:`~pystac.MediaType` for common media types.
+        properties (dict): Optional, additional properties for this asset. This is used
+            by extensions as a way to serialize and deserialize properties on asset
+            object JSON.
+        owner: The Item or Collection this asset belongs to, or None if it has no owner.
+    """
+
+    def __init__(
+        self,
+        href: str,
+        title: Optional[str] = None,
+        description: Optional[str] = None,
+        media_type: Optional[str] = None,
+        roles: Optional[List[str]] = None,
+        properties: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self.href = href
+        self.title = title
+        self.description = description
+        self.media_type = media_type
+        self.roles = roles
+
+        if properties is not None:
+            self.properties = properties
+        else:
+            self.properties = {}
+
+        # The Item which owns this Asset.
+        self.owner: Optional[Union[ps.Item, ps.Collection]] = None
+
+    def set_owner(self, obj: Union["Collection_Type", "Item_Type"]) -> None:
```

Comment:
It's not OK - it will cause circular dependencies (explained a bit more in comment below)

Reply:
Now that I've read up on this more, we eventually won't have to do this, but we are indeed stuck with strings.  I need 3.6 inside google and 3.6 is being checked for CI.  But do see the following for a better future

https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class

https://www.python.org/dev/peps/pep-0563/ will let us do:

```
from __future__ import annotations
```

Reply:
Not super relevant here I don't think, but in case it's interesting to you as well, there was an issue with PEP 563 that caused it's delayed release in 3.10 around libraries like Pydantic using type hints in a way that the postponed evaluation of annotations would disallow https://github.com/samuelcolvin/pydantic/issues/2678. We're not using type hints at runtime in the same way, but interesting nonetheless